File: src/ops/plan.ts
/** wsc plan — classify and dispatch to the right op. */

import { withCall } from "../audit.js";
import { getRouter, type RouteDecision } from "../routing.js";
import * as crawl from "./crawl.js";
import * as discover from "./discover.js";
import * as docs from "./docs.js";
import * as fetchOp from "./fetch.js";
import * as search from "./search.js";

export interface PlanOptions {
  prefer?: "fast" | "deep";
  budgetOverride?: number | null;
  routerName?: "rule" | "llm";
  correlationId?: string;
  noReceipt?: boolean;
}

function shellQuote(s: string): string {
  if (/^[\w\.\/:%@\-]+$/.test(s)) return s;
  return `'${s.replace(/'/g, "'\\''")}'`;
}

export function explain(query: string, opts: PlanOptions = {}): Record<string, unknown> {
  const router = getRouter(opts.routerName ?? "rule");
  const decision = router.classify(query, { prefer: opts.prefer, budgetOverride: opts.budgetOverride });
  return {
    ok: true,
    operation: "plan.explain",
    query,
    decision,
    would_run: `wsc ${decision.recommended_op} ${shellQuote(query)}`,
    returncode: 0,
  };
}

export async function run(query: string, opts: PlanOptions = {}): Promise<Record<string, unknown>> {
  const router = getRouter(opts.routerName ?? "rule");
  const decision: RouteDecision = router.classify(query, {
    prefer: opts.prefer,
    budgetOverride: opts.budgetOverride,
  });
  const correlationId = opts.correlationId ?? process.env.WSC_CORRELATION_ID;

  let subPayload: Record<string, unknown>;
  switch (decision.recommended_op) {
    case "docs":
      subPayload = (await docs.run(query, { correlationId, noReceipt: opts.noReceipt })) as unknown as Record<
        string,
        unknown
      >;
      break;
    case "discover":
      subPayload = await discover.run(query, { correlationId, noReceipt: opts.noReceipt });
      break;
    case "fetch":
      subPayload = await fetchOp.run(query, { correlationId, noReceipt: opts.noReceipt });
      break;
    case "crawl":
      subPayload = await crawl.run(query, { correlationId, noReceipt: opts.noReceipt });
      break;
    case "search":
      subPayload = await search.run(query, { correlationId, noReceipt: opts.noReceipt });
      break;
    default:
      subPayload = {
        ok: false,
        operation: decision.recommended_op,
        error: `unknown op: ${decision.recommended_op}`,
        returncode: 2,
      };
  }

  await withCall(
    "plan",
    { provider: decision.recommended_provider, correlationId, noReceipt: opts.noReceipt },
    async (receipt) => {
      receipt.route_decision = decision;
      receipt.dispatched_op = decision.recommended_op;
      receipt.sub_status = (subPayload.status as string) ?? (subPayload.ok ? "ok" : "error");
    },
  );

  return {
    ok: !!subPayload.ok,
    operation: "plan",
    query,
    decision,
    dispatched_op: decision.recommended_op,
    result: subPayload,
    returncode: Number(subPayload.returncode ?? 0),
  };
}

File: cli/src/wsc/ops/plan.py
"""wsc plan — classify a query and dispatch to the right op.

In v0.1 the dispatcher only handles the unambiguous routing decisions made by
``RuleRouter`` — there's no "auto cross-validation" yet (that's v0.2 once the
content cache and multi-source dedup land). The receipt embeds the full
``RouteDecision`` so downstream consumers (Hermit) can replay the choice.
"""

from __future__ import annotations

import os
import shlex
from typing import Any, Dict, Optional

from wsc import audit, routing
from wsc.ops import crawl, discover, docs, fetch, search


def explain(
    query: str,
    *,
    prefer: Optional[str] = None,
    budget_override: Optional[int] = None,
    router_name: str = "rule",
) -> Dict[str, Any]:
    router = routing.get_router(router_name)
    decision = router.classify(query, context={"prefer": prefer, "budget_override": budget_override})
    return {
        "ok": True,
        "operation": "plan.explain",
        "query": query,
        "decision": decision.to_dict(),
        "would_run": f"wsc {decision.recommended_op} {shlex.quote(query)}",
        "returncode": 0,
    }


def run(
    query: str,
    *,
    prefer: Optional[str] = None,
    budget_override: Optional[int] = None,
    router_name: str = "rule",
    correlation_id: Optional[str] = None,
    no_receipt: bool = False,
) -> Dict[str, Any]:
    router = routing.get_router(router_name)
    decision = router.classify(query, context={"prefer": prefer, "budget_override": budget_override})

    correlation = correlation_id or os.environ.get("WSC_CORRELATION_ID")

    op = decision.recommended_op
    sub_payload: Dict[str, Any]
    if op == "docs":
        sub_payload = docs.run(query, correlation_id=correlation, no_receipt=no_receipt)
    elif op == "discover":
        sub_payload = discover.run(query, correlation_id=correlation, no_receipt=no_receipt)
    elif op == "fetch":
        sub_payload = fetch.run(query, correlation_id=correlation, no_receipt=no_receipt)
    elif op == "crawl":
        sub_payload = crawl.run(query, correlation_id=correlation, no_receipt=no_receipt)
    elif op == "search":
        sub_payload = search.run(query, correlation_id=correlation, no_receipt=no_receipt)
    else:
        sub_payload = {"ok": False, "operation": op, "error": f"unknown op: {op}", "returncode": 2}

    # Wrap with the routing decision and a top-level "plan" receipt so the
    # plan call itself appears in the audit log alongside the dispatched op.
    with audit.start_call("plan", provider=decision.recommended_provider, correlation_id=correlation, no_receipt=no_receipt) as receipt:
        receipt["route_decision"] = decision.to_dict()
        receipt["dispatched_op"] = op
        receipt["sub_status"] = sub_payload.get("status") or ("ok" if sub_payload.get("ok") else "error")

    return {
        "ok": sub_payload.get("ok", False),
        "operation": "plan",
        "query": query,
        "decision": decision.to_dict(),
        "dispatched_op": op,
        "result": sub_payload,
        "returncode": int(sub_payload.get("returncode", 0)),
    }
